### PR TITLE
v0 terraform infs

### PR DESCRIPTION
**Pull request by peterdudfield**
# Pull Request

## Description

Review of first v1 of nowcasting inferstructure

enviornments:
- development

modules:
- vpc
- rds
- ecs
- s3
- services - api
- services - nwp

## How Has This Been Tested?

ran terrafrom apply and it spins up

- [ ] No
- [ ] Yes

## Checklist:

- [ ] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/nowcasting/blob/main/coding_style.md)
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: .github/workflows/terraform-validate.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+name: Validate Terraform
```
Comment (peterdudfield) [inline .github/workflows/terraform-validate.yaml:1]: add comments what this does

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+default_language_version:
+  python: python3.9
```
Comment (peterdudfield) [inline .pre-commit-config.yaml:2]: probably need to change this

File: terraform/development/backend.tf [hunk 3]
```diff
@@ -0,0 +1,8 @@
+terraform {
```
Comment (peterdudfield) [inline terraform/development/backend.tf:1]: add comment and issue about moving to terraform cloud

File: terraform/modules/database/README.md [hunk 4]
```diff
@@ -0,0 +1,4 @@
+This module currently makes
+- RDS postgres database
```
Comment (peterdudfield) [inline terraform/modules/database/README.md:2]: maybe have link to AWS?

File: terraform/modules/database/iam.tf [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# IAM policy for reading reds secret
```
Comment (peterdudfield) [inline terraform/modules/database/iam.tf:1]: typo

File: terraform/modules/database/rds.tf [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# RDS postgres database
+
+resource "aws_db_instance" "DB-forecast" {
+  allocated_storage          = 10
```
Comment (peterdudfield) [inline terraform/modules/database/rds.tf:4]: add issue where storage automatically expands

File: terraform/modules/ecs/ecs.tf [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# Make ECS cluster
+# useful blog - https://engineering.finleap.com/posts/2020-02-20-ecs-fargate-terraform/
+resource "aws_ecs_cluster" "main" {
+  name = "ECS-cluster-${var.environment}"
```
Comment (peterdudfield) [inline terraform/modules/ecs/ecs.tf:4]: should this be renammined to nowcasting-development

File: terraform/modules/s3/s3.tf [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  acl    = "private"
+
+    lifecycle_rule {
+      id = "remove_old_files"
+      enabled = true
+
+      prefix = "data/"
+
+      expiration {
+          days = 180
```
Comment (peterdudfield) [inline terraform/modules/s3/s3.tf:27]: change to 30 days

File: terraform/modules/services/api/docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,9 @@
+version: "3"
+
+services:
+  api:
+    image: openclimatefix/nowcasting_api:0.1.5
```
Comment (peterdudfield) [inline terraform/modules/services/api/docker-compose.yml:5]: add comment about making '0.1.5' a variable,

File: terraform/modules/services/nwp/ecs.tf [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        "options" : {
+          "awslogs-group" : var.log-group-name,
+          "awslogs-region" : var.region,
+          "awslogs-stream-prefix" : "streaming"
+        }
+      }
+    }
+  ])
+
+  #  volume {
```
Comment (peterdudfield) [inline terraform/modules/services/nwp/ecs.tf:50]: remove, but add comment that maybe we need to add a volume

File: terraform/modules/services/nwp/iam.tf [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+   }
+ ]
+}
+EOF
+}
+
+resource "aws_iam_policy" "nwp-secret-read" {
+  name        = "nwp-secret-read"
+  path        = "/consumer/nwp/"
+  description = "Policy to allow read access to NWP secret"
```
Comment (peterdudfield) [inline terraform/modules/services/nwp/iam.tf:28]: add comment that NWP secret holds the API key

File: terraform/modules/database/security_groups.tf [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+  vpc_id = var.vpc_id
+}
+
+# Inbound Security rule
+resource "aws_security_group_rule" "rds-inbound" {
+  type              = "ingress"
+  from_port         = 5432
+  to_port           = 5432
+  protocol          = "tcp"
+  cidr_blocks       = ["0.0.0.0/0"]
```
Comment (flowirtz) [inline terraform/modules/database/security_groups.tf:14]: This means that we allow anyone to access our db.
Is that intentional?

I'd prefer to lock it down to eg a certain SG via `source_ security_group_id`

**Comment by peterdudfield** [comment]
Thanks very much @flowirtz for looking at these, Ill have a look through them now

File: terraform/modules/s3/s3.tf (hunk 8 above)
Comment (peterdudfield) [reply terraform/modules/s3/s3.tf]: The idea is to just keep 30 days for inference. The backtesting is down with a different pipeline at the moment. Can easily change this

File: terraform/development/backend.tf [hunk 13]
```diff
@@ -0,0 +1,8 @@
...
+  backend "s3" {
+    encrypt = true
+    bucket  = "nowcasting-terraform"
+    key     = "terraform.tfstate"
+    region  = "eu-west-2"
```
[the remark this answers is not in the document]
Comment (peterdudfield) [reply terraform/development/backend.tf:6]: Sounds good, ill move to Ireland

File: terraform/modules/database/rds.tf [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+  publicly_accessible        = false
+  vpc_security_group_ids     = [aws_security_group.rds-postgres-sg.id]
+  ca_cert_identifier         = "rds-ca-2019"
+  backup_window              = "00:00-00:30"
+  db_subnet_group_name       = var.db_subnet_group.name # update name with private/public
+  auto_minor_version_upgrade = true
+  performance_insights_enabled = true
+
+  tags = {
+    Name        = "${var.environment}-rds"
```
[the remark this answers is not in the document]
Comment (peterdudfield) [reply terraform/modules/database/rds.tf:22]: It seems ok, but I probably do want to go thorugh and do the Tags properly at some point. Ill add an issue

File: terraform/modules/database/secrets.tf [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+  override_special = "%"
+}
+
+
+# Now create secret and secret versions for database main account
+resource "aws_secretsmanager_secret" "DB-forecast-secret" {
+  name = "RDS/forecast/main"
+  # Once the secret is deleted, we cant get it back.
+  # If we don't do this, then a new secret can be made with the same name until the recovery window is over
+  recovery_window_in_days = 0
```
[the remark this answers is not in the document]
Comment (peterdudfield) [reply terraform/modules/database/secrets.tf:17]: hmm, yea, its a tricky one. The annoying thing if we keep it, we cant destory the stack and re build it and it says the secret is (kinda) still there. This is really annoying when development. Perhaps for production we can think of something a bit different

File: terraform/modules/services/nwp/cloudwatch.tf [hunk 16]
```diff
@@ -0,0 +1,12 @@
+# set up cloudwatch log group
+
+resource "aws_cloudwatch_log_group" "nwp" {
+  name = var.log-group-name
+
+  retention_in_days = 7
+
+  tags = {
+    Environment = var.environment
+    Application = "nowcasting"
```
[the remark this answers is not in the document]
Comment (peterdudfield) [reply terraform/modules/services/nwp/cloudwatch.tf:10]: Yea, I was thinking we just have one aws acoount for nowcasting, and a different one for different ones, but still, we should put that tag in. Ive added it to https://github.com/openclimatefix/nowcasting_infrastructure/issues/24

File: terraform/modules/networking/main.tf [hunk 17]
```diff
@@ -46,58 +57,65 @@ resource "aws_subnet" "public_subnet" {
...
 }
 /*==== VPC's Default Security Group ======*/
 resource "aws_security_group" "default" {
-  name        = "${var.environment}-default-sg"
+  name        = "nowcasting-${var.environment}-default-sg"
   description = "Default security group to allow inbound/outbound from the VPC"
-  vpc_id      = "${aws_vpc.vpc.id}"
+  vpc_id      = aws_vpc.vpc.id
   depends_on  = [aws_vpc.vpc]
   ingress {
```
[the remark this answers is not in the document]
Comment (peterdudfield) [reply terraform/modules/networking/main.tf:120]: Yea I dont know, maybe for reaching external APIs?

File: terraform/modules/database/security_groups.tf (hunk 12 above)
Comment (peterdudfield) [reply terraform/modules/database/security_groups.tf:14]: That certainly isnt what i meant. It does have a secruity group below, so maybe it is on that. 

I'll https://github.com/openclimatefix/nowcasting_infrastructure/issues/25

File: terraform/modules/database/secrets.tf (hunk 15 above)
[the remark this answers is not in the document]
Comment (peterdudfield) [reply terraform/modules/database/secrets.tf:17]: Ill leave this for the moment

**Comment by peterdudfield** [comment]
@all-contributors please add @flowirtz for review

**Pull request merged**